### ISSUE TITLE
Incorrect error on legitimate type overlap

### ISSUE THREAD

**Issue by millsp** [microsoft/TypeScript#41402]
**TypeScript Version:**  3.7.x-dev.201xxxxx

**Search Terms:**

This condition will always return 'false' since the types 'X' and 'Y' have no overlap.

**Code**

```ts
function fn<T>(t: T) {
    return t === '' // error
}

fn('') // but here we pass it
```

https://www.typescriptlang.org/play?jsx=0#code/GYVwdgxgLglg9mABMMAeAKgPgBRQFyLoCUiA3gFCJWIBOAplCDUlIgLweIDkXiA9H0R0aNODXIBfcuRTYeJAYgDuYgNYBnIA

**Expected behavior:**

```ts
function fn<T extends unknown>(t: T) {
    return t === '' // works
}

fn('') // works
```

https://www.typescriptlang.org/play?jsx=0#code/GYVwdgxgLglg9mABMMAeAKogpgDylsAEwGdFwBrMOAdzAD4AKKALkXQEpEBvAKEX8QAnLFBCCkURAF4ZiAORzEAeiWJqcQeWI8Avjx4oGCzirUatQA

**Actual behavior:**

> `This condition will always return 'false' since the types 'T' and 'string' have no overlap.`

**Related issues**
https://github.com/microsoft/TypeScript/issues/27910

**Comment by RyanCavanaugh** [comment]
This is intentionally an error, but the wording needs work

**Comment by millsp** [comment]
Hey @RyanCavanaugh 

Why would this be an error? It's simple, valid JS:

```ts
function fn<T>(t: T) {
    return t === '' // error
}

fn('') // but here we pass it
```

**Comment by RyanCavanaugh** [comment]
https://stackoverflow.com/questions/41750390/what-does-all-legal-javascript-is-legal-typescript-mean

**Comment by millsp** [comment]
I actually expected an explanation why this is OK

```ts
function fn<T extends unknown>(t: T) {
    return t === '' // works
}

fn('') // works
```

and works perfectly, but not the equivalent

```ts
function fn<T>(t: T) {
    return t === '' // fails
}

fn('') // works
```

@RyanCavanaugh would you be so kind to explain, please :)

**Comment by RyanCavanaugh** [comment]
A generic type parameter is comparable to the type of its constraint if a constraint was explicitly declared

**Comment by millsp** [comment]
Should we build a library to replace our comparison operators, to make this simple use-case work?

```ts
const equals = <A>(thing: unknown, other: A): thing is A => {
    return thing === other;
};

function fn<T>(t: T) {
    if (equals(t, '')) {
        // t: T & string
    } // works
}

fn(''); // works
```

Why is the `equals` type guard working but not the basic `===`?

**Comment by MartinJohns** [comment]
> Why is the `equals` type guard working but not the basic `===`?

`unknown` and `any` overlap with the generic type, so you're allowed to compare them. `T` is always assignable to `unknown` and `any`.

The same would apply without the additional type-guard:

    function fn<T>(t: T) {
        const val: unknown = '';
        if (t === val) {
            t // t is T
        }
    }

Just of course the type information `string` is lost and `t` is `T` instead of `T & string`.

**Comment by millsp** [comment]
> Just of course the type information `string` is lost and `t` is `T` instead of `T & string`

That's exactly what I'm complaining about.

> The same would apply without the additional type-guard

I get that a generic `T` intersected with `any` or `unknown` yields `T`.

**I'm pointing out the fact that we could safely narrow `T` with the type that sits on the `rhs` of the comparison operator.**

But @RyanCavanaugh seems to say that it's not possible, and I would like to know why should we have such a limitation. 

Another example:

```ts
function fn<A, B>(a: A, b: B) {
    if (a === b) { // fails here
		// ...
    } else {	
		// ...
    }
}
```

```ts
const equals = <A>(thing: unknown, other: A): thing is A => {
    return thing === other;
};

function fn<A, B>(a: A, b: B) {
    if (equals(a, b)) { // works
        // a: A & B
    } else {
        // a: A
    }
}
```
